common/determined_common/_logging.py

```diff
@@ -1,16 +1,207 @@
+import contextlib
+import json
 import logging
 import sys
+import threading
+import time
+from typing import Any, Callable, Dict, Generator, Optional, Union
 
 
-def set_logger(debug_enabled: bool) -> None:
-    root = logging.getLogger()
-    root.setLevel(logging.DEBUG if debug_enabled else logging.INFO)
+class _Loggers:
+    def __init__(self) -> None:
+        self._storage = logging.getLogger("det.storage")
+        self._harness = logging.getLogger("det.harness")
+        self._resources = logging.getLogger("det.resources")
+        self._all_loggers = (self._storage, self._harness, self._resources)
 
-    for hdlr in root.handlers:
-        root.removeHandler(hdlr)
+    @property
+    def storage(self) -> logging.Logger:
+        return self._storage
 
-    handler = logging.StreamHandler(sys.stdout)
-    handler.setLevel(logging.DEBUG if debug_enabled else logging.INFO)
-    formatter = logging.Formatter("%(levelname)s: %(message)s")
-    handler.setFormatter(formatter)
-    root.addHandler(handler)
+    @property
+    def harness(self) -> logging.Logger:
+        return self._harness
+
+    @property
+    def resources(self) -> logging.Logger:
+        return self._resources
+
+
+# Loggers are normally global in python, and it would be a bit unexpected to behave differently.
+log = _Loggers()
+
+
+class _ProfilingThread(threading.Thread):
+    """
+    Call a profiling function periodically on a separate thread.
+    """
+
+    def __init__(self, period: float, profile_fn: Callable) -> None:
+        self.period = period
+        self.profile_fn = profile_fn
+        self.quitting = False
+        self.cond = threading.Condition()
+        super().__init__()
+
+    def run(self) -> None:
+        with self.cond:
+            while True:
+                self.cond.wait(timeout=self.period)
+                if self.quitting:
+                    break
+                self.profile_fn()
+
+    def quit(self) -> None:
+        self.quitting = True
```

Comment:
there's no need to; it would not affect the behavior of the system at all since it is a bool that is only changed once in the course of its lifetime.  I have a tiny preference for leaving it out of the critical section since it doesn't literally have to be there.

But since there's a relevant critical section just below, I could move it there if you cared much at all.

Reply:
(answering below, I think I moved this class to the `harness` package and forgot to delete it here)